src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -120,138 +149,162 @@ static void *EventProcessingThread(void *data) {
+    ConstFSEventStreamRef streamRef,
+    void *clientCallBackInfo,
+    size_t numEvents,
+    void *eventPaths,
+    const FSEventStreamEventFlags eventFlags[],
+    const FSEventStreamEventId eventIds[]
+) {
+    Server *server = (Server*) clientCallBackInfo;
+    server->handleEvents(numEvents, (char **) eventPaths, eventFlags, eventIds);
+}
 
-    if (threadLoop != NULL) {
-        CFRunLoopStop(threadLoop);
+void Server::handleEvents(
+    size_t numEvents,
+    char **eventPaths,
+    const FSEventStreamEventFlags eventFlags[],
+    const FSEventStreamEventId eventIds[]
+) {
+    if (invalidStateDetected) {
+        // TODO Handle this better
+        return;
     }
 
-    if (watcherThread != NULL) {
-        pthread_join(watcherThread, NULL);
+    JNIEnv* env = getThreadEnv();
+
+    for (int i = 0; i < numEvents; i++) {
+        handleEvent(env, eventPaths[i], eventFlags[i]);
     }
+}
 
-    if (rootsToWatch != NULL) {
-        for (int i = 0; i < CFArrayGetCount(rootsToWatch); i++) {
-            const void *value = CFArrayGetValueAtIndex(rootsToWatch, i);
-            CFRelease(value);
+void Server::handleEvent(JNIEnv *env, char* path, FSEventStreamEventFlags flags) {
+    log_fine(env, "Event flags: 0x%x for %s", flags, path);
+
+    jint type;
+    if (IS_SET(flags, kFSEventStreamEventFlagHistoryDone)) {
+        return;
+    } else if (IS_ANY_SET(flags,
+            kFSEventStreamEventFlagRootChanged
+            | kFSEventStreamEventFlagMount
+            | kFSEventStreamEventFlagUnmount
+            | kFSEventStreamEventFlagMustScanSubDirs)) {
+        type = FILE_EVENT_INVALIDATE;
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemRenamed)) {
+        if (IS_SET(flags, kFSEventStreamEventFlagItemCreated)) {
+            type = FILE_EVENT_REMOVED;
+        } else {
+            type = FILE_EVENT_CREATED;
         }
-        // TODO Can we release these earlier?
-        CFRelease(rootsToWatch);
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemModified)) {
+        type = FILE_EVENT_MODIFIED;
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemRemoved)) {
+        type = FILE_EVENT_REMOVED;
+    } else if (IS_ANY_SET(flags,
+            kFSEventStreamEventFlagItemInodeMetaMod // file locked
+            | kFSEventStreamEventFlagItemFinderInfoMod
+            | kFSEventStreamEventFlagItemChangeOwner
+            | kFSEventStreamEventFlagItemXattrMod)) {
+        type = FILE_EVENT_MODIFIED;
+    } else if (IS_SET(flags, kFSEventStreamEventFlagItemCreated)) {
+        type = FILE_EVENT_CREATED;
+    } else {
+        log_warning(env, "Unknown event 0x%x for %s", flags, path);
+        type = FILE_EVENT_UNKNOWN;
     }
 
-    if (watcherStream != NULL) {
-        FSEventStreamStop(watcherStream);
-        FSEventStreamInvalidate(watcherStream);
-        FSEventStreamRelease(watcherStream);
-        // TODO: consider using FSEventStreamFlushSync to flush all pending events.
+    // TODO What does this do?
```

Comment:
Looks like it selects the first line from a multiline string by adding the terminator where the first line break is.

Reply:
Right. I guess the comment should have said "Why is this necessary?". Any ideas for that? We might as well remove this code and try to run the tests. I'm planning to do it in a separate PR.

Reply:
👍